### PR TITLE
 Bug 1429882: deprecations fix for openURL and String.characters

### PR DESCRIPTION
**Pull request by garvankeeley**
https://bugzilla.mozilla.org/show_bug.cgi?id=1429882

### CODE REVIEW

File: Client/Extensions/NSURLExtensionsMailTo.swift [hunk 1]
```diff
@@ -26,7 +26,7 @@ public extension URL {
         let urlString = absoluteString
 
         // Extract 'to' value
-        let toStart = urlString.characters.index(urlString.startIndex, offsetBy: "mailto:".characters.count)
+        let toStart = urlString.characters.index(urlString.startIndex, offsetBy: "mailto:".count)
```
Comment (justindarc) [inline Client/Extensions/NSURLExtensionsMailTo.swift]: Isn't there still another use of `characters` here?

File: Client/Frontend/Widgets/ActivityStreamHighlightCell.swift [hunk 2]
```diff
@@ -162,7 +162,7 @@ class ActivityStreamHighlightCell: UICollectionViewCell {
         }
 
         self.domainLabel.text = site.tileURL.hostSLD
-        self.titleLabel.text = site.title.characters.count <= 1 ? site.url : site.title
+        self.titleLabel.text = site.title.isEmpty ? site.url : site.title
```
Comment (justindarc) [inline Client/Frontend/Widgets/ActivityStreamHighlightCell.swift]: This looks like a change in logic?

File: Client/Frontend/Widgets/PhotonActionSheet.swift [hunk 3]
```diff
@@ -375,7 +375,7 @@ private class PhotonActionSheetHeaderView: UITableViewHeaderFooterView {
             self.siteImageView.backgroundColor = color
             self.siteImageView.image = self.siteImageView.image?.createScaled(PhotonActionSheetUX.IconSize)
         }
-        self.titleLabel.text = site.title.characters.count <= 1 ? site.url : site.title
+        self.titleLabel.text = site.title.isEmpty ? site.url : site.title
```
Comment (justindarc) [inline Client/Frontend/Widgets/PhotonActionSheet.swift]: This again looks like a change in logic. Its weird logic to begin with though so, I'm not sure if we should keep it.

File: UITests/ToolbarTests.swift [hunk 4]
```diff
@@ -28,7 +28,7 @@ class ToolbarTests: KIFTestCase, UITextFieldDelegate {
         let url = "\(localhostURL)/numberedPage.html?page=1"
 
         // URL without "http://".
-        let displayURL = "\(localhostURL)/numberedPage.html?page=1".substring(from: url.characters.index(url.startIndex, offsetBy: "http://".characters.count))
+        let displayURL = "\(localhostURL)/numberedPage.html?page=1".substring(from: url.characters.index(url.startIndex, offsetBy: "http://".count))
```
Comment (justindarc) [inline UITests/ToolbarTests.swift]: I almost wonder if we shouldn't just hard-code `7` here?

File: XCUITests/FxScreenGraph.swift [hunk 5]
```diff
@@ -791,7 +791,7 @@ extension Navigator where T == FxUserState {
 
     // Open a URL. Will use/re-use the first BrowserTab or NewTabScreen it comes to.
     @available(*, deprecated, message: "use openURL(_ urlString) instead")
-    func openURL(urlString: String) {
+    func openURL(String) {
```
Comment (justindarc) [inline XCUITests/FxScreenGraph.swift]: How does this work?

File: Client/Frontend/Widgets/ActivityStreamHighlightCell.swift (hunk 2 above)
Comment (garvankeeley) [reply Client/Frontend/Widgets/ActivityStreamHighlightCell.swift]: Nice catch, it is an accidental logic change.
Hmm, not sure about the '2 chars is the min title length' logic in that code –I doubt we know for sure that 1 character makes for an invalid title in all languages.
Can I keep this change?

File: Client/Extensions/NSURLExtensionsMailTo.swift (hunk 1 above)
Comment (garvankeeley) [reply Client/Extensions/NSURLExtensionsMailTo.swift]: Yes, I'll search for more usages I missed.

File: UITests/ToolbarTests.swift (hunk 4 above)
Comment (garvankeeley) [reply UITests/ToolbarTests.swift]: I like the "string".len syntax for clarity/safety

File: XCUITests/FxScreenGraph.swift (hunk 5 above)
Comment (garvankeeley) [reply XCUITests/FxScreenGraph.swift]: Right, I should have removed that entirely, will do

**Comment by garvankeeley** [comment]
Oh wow, I missed a ton of .characters calls, new PR coming now...

File: Client/Frontend/Browser/Punycode.swift [hunk 6]
```diff
@@ -11,7 +11,7 @@ private let initialBias = 72
 private let initialN: Int = 128 // 0x80
 private let delimiter: Character = "-"; // '\x2D'
 private let prefixPunycode = "xn--"
-private let asciiPunycode = [Character]("abcdefghijklmnopqrstuvwxyz0123456789".characters)
+private let asciiPunycode = Array("abcdefghijklmnopqrstuvwxyz0123456789")
```
Comment (garvankeeley) [inline Client/Frontend/Browser/Punycode.swift]: A cool swift way to get a string as an array

File: Client/Frontend/Widgets/ActivityStreamHighlightCell.swift (hunk 2 above)
Comment (justindarc) [reply Client/Frontend/Widgets/ActivityStreamHighlightCell.swift]: Yeah. This logic makes no sense to me. Let's just keep what you have.

File: Client/Frontend/Widgets/PhotonActionSheet.swift (hunk 3 above)
Comment (justindarc) [reply Client/Frontend/Widgets/PhotonActionSheet.swift]: Disregard this comment. Let's keep your new logic.

**Pull request merged**